blivet/iscsi.py

```diff
@@ -330,95 +324,64 @@ def discover(self, ipaddr, port="3260", username=None, password=None,
-            # using signals internally which are send to bad thread
-            (con_recv, con_write) = Pipe(False)
-            p = Process(target=_call_discover_targets, args=(con_write,
-                                                             con_recv,
-                                                             ipaddr,
-                                                             port,
-                                                             authinfo, ))
-            p.start()
-
-            # Close the writing end of the pipe in the parent
-            con_write.close()
-
-            with con_recv:
-                try:
-                    (ok, data) = con_recv.recv()
-                    if not ok:
-                        log.debug("iSCSI: exception raised when "
-                                  "discover_sendtargets process called: %s",
-                                  str(data))
-                except EOFError:
-                    ok = False
-                    log.error("iSCSI: can't receive response from "
-                              "_call_discover_targets")
-
-            p.join()
-
-            # convert dictionary back to iscsi nodes object
-            if ok:
-                found_nodes = []
-                for node in data:
-                    found_nodes.append(libiscsi.node(**node))
-            else:
-                return []
-
-            self.discovered_targets[(ipaddr, port)] = []
+            auth_info = dict()
+            if username:
+                auth_info["username"] = GLib.Variant("s", username)
+            if password:
+                auth_info["password"] = GLib.Variant("s", password)
+            if r_username:
+                auth_info["r_username"] = GLib.Variant("s", r_username)
+            if r_password:
+                auth_info["r_password"] = GLib.Variant("s", r_password)
+
+            args = GLib.Variant("(sqa{sv}", tuple(ipaddr, port, auth_info))
+            nodes, _n_nodes = self._call_initiator_method("DiscoverSendTargets", args)
+
+            found_nodes = _to_node_infos(nodes)
+            t_info = TargetInfo(ipaddr, port)
+            self.discovered_targets[t_info] = []
             for node in found_nodes:
-                self.discovered_targets[(ipaddr, port)].append([node, False])
+                self.discovered_targets[t_info].append(LoginInfo(node, False))
                 log.debug("discovered iSCSI node: %s", node.name)
 
         # only return the nodes we are not logged into yet
-        return [node for (node, logged_in) in
-                self.discovered_targets[(ipaddr, port)]
-                if not logged_in]
+        return [info.node for info in self.discovered_targets[TargetInfo(ipaddr, port)]
+                if not info.logged_in]
 
     def log_into_node(self, node, username=None, password=None,
                       r_username=None, r_password=None):
-        """
-        Raises IOError.
-        """
```

Comment:
It would be nice if you could instead of only removing this docstring also add new one.